network/p2p/scoring/registry.go

```diff
@@ -401,12 +417,24 @@ func (r *GossipSubAppSpecificScoreRegistry) subscriptionPenalty(pid peer.ID, flo
 	return 0
 }
 
+// duplicateMessagesPenalty returns the duplicate message penalty for a peer.
+func (r *GossipSubAppSpecificScoreRegistry) duplicateMessagesPenalty(pid peer.ID) float64 {
+	duplicateMessageCount := r.getDuplicateMessageCount(pid)
+	if duplicateMessageCount != 0 {
+		duplicateMessagePenalty := duplicateMessageCount * DefaultDuplicateMessagePenalty
+		if duplicateMessagePenalty < MaxAppSpecificPenalty {
+			return MaxAppSpecificPenalty
+		}
+		return duplicateMessagePenalty
```

Comment:
Please collect this in a histogram.

Reply:
Will using the peerID as a label cause a strain in our prometheus instances since we can have any number of peerIDs? We currently have a log message that has all penalty information for a peer when the app specific score is computed, will the histogram give us some different insight?

Reply:
Correct, but we require a straightforward `Histogram` rather than a `HistogramVec`, hence, no `label` would be needed. We only need to track all the returned `duplicateMessagePenalties` as separate data points, which a `Histogram` empowers. This enables us to perform queries such as determining the 0.95 percentile for duplicate message penalties across all nodes. Such queries are crucial for debugging incidents that involve a drastic decrease in a node's overall score, helping us identify the contributing factors. Therefore, I would suggest please to use a simple `Histogram` and let it observe any returned values in here.

I would also recommend having a separate `Histogram` tracking the `duplicateMessageCount` as well. These two histograms are one for the count and one for the penalty will provide a close lense in how the node behaves w.r.t to the duplicate messages it receives, as well as the health of decaying mechanism. 

>Will using the peerID as a label cause a strain in our prometheus instances since we can have any number of peerIDs? 

Since we are using a `Histogram` than a `HistogramVec`, there is no cardinality beyond the number of buckets that we set for the histogram. In other words, as we don't record the `peerId`s when observing the values, there won't be any scalability issue with the metrics cardinality relative to the number of peers. 

This is a similar approach we took for [observing the application-specific score](https://github.com/onflow/flow-go/blob/master/module/metrics/gossipsub_score.go#L134) of nodes in a [`Histogram`](https://github.com/onflow/flow-go/blob/master/module/metrics/gossipsub_score.go#L43-L51).

Reply:
https://github.com/onflow/flow-go/pull/5058/commits/8729c4b32088bdd73fb0044982600c2c18605f25